e621ng/e621ng
# EditHistories overhaul & search

### Changes
* [EditHistories] Adds a new `edit_type` column to EditHistory entries. Currently, there's:
  * `original` - `version=1`, nothing special
  * `edit` - any content change
  * `hide`/`unhide` - hiding or unhiding
  * `stick`/`unstick` - sticking and unsticking
  * `mark_warning`/`mark_record`/`mark_ban` - marking for a warning/record/ban
  * `unmark` - removing any mark
* [EditHistoriesController] Adds searching to `/edit_histories` (`versionable_type`, `versionable_id`, `edit_type`, `user_id`, `user_name`, `ip_addr` (admin))
  * `edit_type=original` is excluded, unless searching for a specific edit type (`original` is still not included in the dropdown)
* [Blips/Comments/ForumPosts] Add timestamp for hiding to the "(hidden)" text (`title` attribute, shown when hovering)
  * This does remove the abilitiy of users immediately seeing who hid their comment, though they can still find out by looking at the modactions
  * [Example](https://github.com/e621ng/e621ng/assets/17226394/bc282c9f-06e7-431b-b7ff-c72627428853)
* [EditHistories] Add "Listing" link to "secondary links" to return to index
* [EditHistories#index] Add `|| ""` to subject column to fix `false` spam
* [EditHistories#index] Change "Show" link to actually link directly to that edit (via anchor)
* [EditHistories#show] Add link to "Edits for X #0" header (e.g. "Edits for [Comment #0](http://localhost:3000/comments/0)")
  * This is messy currently, it could potentially be cleaned up if paths can be referenced within models
* [EditHistories#show] Add an `id` to divs (`edit-0`) for linking
* [EditHistories#show] Make time clickable for link, similar to how comments and such behave
* [EditHistories#show] Add numbered paginator
* [WarningNotice] Display timestamp & user inline (when known)
  * [Known Example](https://user-images.githubusercontent.com/17226394/242752573-76a1171f-883d-4309-ac09-eee5eb982cad.png)
  * [Unknown Example](https://user-images.githubusercontent.com/17226394/242752721-c4c6e8c0-42a7-4f55-a6fc-e8117eacda08.png)
  * "Unknown" should only happen on blips marked before this pr, due to `warning_user_id` never being set, and blips not having an `updater_id` we can use for assumptions
* [UserStatistics] Adds link to list forum post/comment edits on user profile

### Tests
* Remove `creator` default from blip factory - this causes mismatches due to `CurrentUser` not being used unless explicitly provided
* Add blip tests - mirrored and trimmed down from comment tests
* Add hide/stick/delete modaction tests to comments & forum posts
* Extensive tests on blips/comments/forum posts to ensure edit histories are being created correctly ([sub]If tests can be shared between files/templated, these could possibly be condensed down)

### General Cleanup
* [EditHistory] Removed `TYPE_MAP`, seems unused
* [ApplicationRecord] Moved creation of `original` histories to `save_original_version` (`save_version` method was over 30 lines)
* [ApplicationRecord] Use `saved_change_to_attribute?` instead of `send`, this won't error if columns don't exist

### Tangentially Related
These are not directly related to edit histories, but including them here can save headaches and possible merge conflicts in the future.
* [UpdateNotice] Changes edited timestamps to always show to Moderators and higher (0f70ecb40a0de4d4c0add4178d9aad311acd171c)
* [UserWarnable] Fixes a bug where `warning_user_id` is never set (6cda2238b30da5c4ba3db163589de1e1d7780ad5)
* [Blips] Adds an `updater_id` field to blips, and adds an update notice to them (d29b239141fb213276daeae6c857beb9616df442)
* [Blips] Add `hide!` & `unhide!` methods to blips (this is the way comments & forum posts are) (8340beab5fcdafecece4ac40c7eaf8ecbbf024ea)
* [Blips] Move modaction generation logic into rails callbacks (8340beab5fcdafecece4ac40c7eaf8ecbbf024ea, db66b3ce55754695a6e3349577ee2caed7dbd063)
* [ForumPosts] Disable creating an edit modaction when hiding forum posts (8693454dea47fd33f10bba97d969b66501a98b8c)
* [ForumPosts] Remove `updater_id != creator_id` check on destroy (66e9b344fd00d3634bce2bb9a8c079abf511efbb)
* [Comments] Always create edit modaction when sticking/unsticking comments (dd4315b10cb6cd800e0c39d7841ebb36e0b0377c)

<details>
<summary><h3>Current Concerns</h3></summary>

The only thing I'm entirely sure needs to changed is the way the non-content edits are displayed within the edit histories. Right now there's just some bold text printed out. This could probably use some different background coloring at the least.
![image](https://github.com/e621ng/e621ng/assets/17226394/b822f856-3543-4c77-af00-e003bd308653)
<hr>
I excluded sticking a comment from creating a modlog action, though this pr adds an edit history entry for sticking & unsticking/
https://github.com/e621ng/e621ng/blob/d1478383b7a99fea115e83ef7f4c5ab0a4d8987c/app/models/comment.rb#L14
</details>



Side tangent, editing (`PATCH /{type}/{id}`) with `is_hidden` doesn't create a "hide"/"unhide" mod action, it only creates an "edited" action. Doing this to other's posts is only available to admins, and not available within normal usage, so it shouldn't be too much of a concern. They will also still get an edit history for whatever has been done.

### TODO
After this is merged, open pull request for including forum topics: subject, locking, hiding, etc.
